File: day6/C/day6.c
#include <stdio.h>
#include <stddef.h>
#include <stdlib.h>
#include <unistd.h>
#include <fcntl.h>

void add_element(char vect[14], char new_ele)
{
    int i;

    i = 0;
    while (i < 13)
    {
        vect[13 - i] = vect[12 - i];
        i++;
    }
    vect[0] = new_ele;
}

int check_vect(char vect[14])
{
    int i;
    int j;

    i = 0;
    while (i < 14)
    {
        j = 0;
        while (j < i)
        {
            if(vect[i] == vect[j])
                return (0);
            j++;
        }
        i++;
    }
    return (1);
}

int find_pos_marker(char* file_name)
{
    int fd;
    char inspector[14] = {0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0};
    int total;
    char char_read;
    fd = open(file_name, O_RDONLY);
    if (fd < 0)
        return (-1);
    total = 0;
    while(1)
    {
        if (read(fd, &char_read, 1) <= 0)
            return (0);
        else
            total++;
        add_element(inspector, char_read);
        if (total > 14 && check_vect(inspector))
            return (total);
    }
}

int main()
{
    printf("marker position: %d\n", find_pos_marker("..\\input.txt"));
    return (0);
}

File: day6/python/day6.py
from collections import Counter

def unique(strng):
    cnt_dict = Counter(strng)
    for count in cnt_dict.values():
        if count != 1:
            return False
    return True


def find_marker_position(file_name, size_marker):
    marker = ""
    tot_chars = 0
    with open(file_name, 'r') as fd:
        while True:
            char = fd.read(1)
            marker += char
            tot_chars += 1
            if len(marker) == size_marker:
                if unique(marker):
                    return tot_chars
                else:
                    marker = marker[1:]
            

if __name__ == "__main__":
    print("marker position:", find_marker_position("..\\input.txt", 4))
    print("marker position:", find_marker_position("..\\input.txt", 14))